hazelcast/src/hazelcast/client/connection/InSelector.cpp

```diff
@@ -66,7 +66,7 @@ namespace hazelcast {
                             sleepingSocket->receive(&wakeUpSignal, sizeof(int));
                         } else {
                             boost::shared_ptr<Connection> conn = connectionManager.getConnectionIfAvailable(fd);
-                            if (conn.get() != NULL) {
+                            if (conn.get() != NULL && conn->live) {
```

Comment:
Current situation: https://github.com/hazelcast/hazelcast-cpp-client/blob/master/hazelcast/src/hazelcast/client/connection/Connection.cpp#L114  removal is after socket.close.  Yes, reversing the order may also help decrease this possibility, but what about already scheduled sockets waiting on select call. May be I can do this: In addition to this ordering, also trigger IOSelector wakeup after the removal from the list.